### ISSUE TITLE
activate outputs NaN/Infinity

### ISSUE THREAD

**Issue by zammbi** [wagenaartje/neataptic#17]
I'm inputting 5 values to the Neat activate function and I'm randomly getting NaN/Infinity on the latest version (1.1.7). 
My inputs are a vaild number of 0-1.  Eg  
> [0, 1E-08, 0.01336913588, 4E-08, 1.8E-07]

```
neat = new Neat(5, 1, null, {
            mutation: [
            Methods.Mutation.ADD_NODE,
            Methods.Mutation.SUB_NODE,
            Methods.Mutation.ADD_CONN,
            Methods.Mutation.SUB_CONN,
            Methods.Mutation.MOD_WEIGHT,
            Methods.Mutation.MOD_BIAS,
            Methods.Mutation.MOD_ACTIVATION,
            Methods.Mutation.ADD_GATE,
            Methods.Mutation.SUB_GATE,
            Methods.Mutation.ADD_SELF_CONN,
            Methods.Mutation.SUB_SELF_CONN,
            Methods.Mutation.ADD_BACK_CONN,
            Methods.Mutation.SUB_BACK_CONN
        ],
        
        mutationRate: 0.4,
        popsize: 500,
        elitism: 50
    });

....

var output = this.brain.activate(input);
```

**Comment by wagenaartje** [comment]
Related https://github.com/wagenaartje/neataptic/issues/13. I expect it to be related to an Absolute node with a selfconnection, which will eternally increase its value.

Could you try to perform the following:

    Methods.Mutation.MOD_ACTIVATION.allowed =  [
      Activation.LOGISTIC,
      Activation.TANH,
      Activation.RELU,
      Activation.IDENTITY,
      Activation.STEP,
      Activation.SOFTSIGN,
      Activation.SINUSOID,
      Activation.GAUSSIAN,
      Activation.BENT_IDENTITY,
      Activation.BIPOLAR,
      Activation.BIPOLAR_SIGMOID,
      Activation.HARD_TANH,
      //Activation.ABSOLUTE
    ];

This disables the absolute activation function, if you're still getting `NaN/Infinity`, could you upload the JSON of the network so I can debug it.

**Comment by zammbi** [comment]
Yes, it still getting NaN/Infinity. Json of the network attached. Let me if you need more info.

Inputs that caused Nan: [0, 1E-08, 0.01339393867, 5E-08, 1E-08]

[Nan network.txt](https://github.com/wagenaartje/neataptic/files/1004266/Nan.network.txt)

**Comment by wagenaartje** [comment]
I'll check it out asap.

I just released a new update introducing an `evolve()` function for networks, so i'm going to edit the wiki first and then i'll look.

**Comment by wagenaartje** [comment]
@zammbi I just tried to debug. Did you use `Network.toJSON()` to export it? You just gave me an array of nodes.

**Comment by zammbi** [comment]
Oh whoops  I used export. I'll get you a copy tonight.

**Comment by zammbi** [comment]
This was causing a NaN:
`{"nodes":[{"bias":0,"type":"input","squash":"LOGISTIC","index":"0"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"1"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"2"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"3"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"4"},{"bias":0.9477200467944603,"type":"hidden","squash":"LOGISTIC","index":"5"},{"bias":-0.05208729288807064,"type":"hidden","squash":"IDENTITY","index":"6"},{"bias":-9.572427436964702,"type":"output","squash":"BIPOLAR_SIGMOID","index":"7"}],"connections":[{"weight":1,"from":"5","to":"5","gater":null},{"weight":1,"from":"6","to":"6","gater":null},{"weight":1,"from":"7","to":"7","gater":null},{"weight":0.8082973635800295,"from":0,"to":5,"gater":null},{"weight":0.9711083233377172,"from":1,"to":7,"gater":null},{"weight":-0.03469677774100703,"from":3,"to":5,"gater":null},{"weight":0.05784720186679079,"from":2,"to":7,"gater":null},{"weight":-0.03739191559004472,"from":7,"to":5,"gater":null},{"weight":1.2546905487590903,"from":5,"to":7,"gater":null},{"weight":0.38875081275612705,"from":4,"to":6,"gater":null},{"weight":-0.6437443458225839,"from":6,"to":5,"gater":null},{"weight":-0.034602202855597716,"from":4,"to":5,"gater":null},{"weight":0.028375978723404804,"from":7,"to":6,"gater":null}],"input":5,"output":1}`

**Comment by zammbi** [comment]
This is causing an

**Comment by zammbi** [comment]
This is causing an "Infinity".

`{"nodes":[{"bias":0,"type":"input","squash":"LOGISTIC","index":"0"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"1"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"2"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"3"},{"bias":0,"type":"input","squash":"LOGISTIC","index":"4"},{"bias":-0.7003282260440697,"type":"hidden","squash":"BENT_IDENTITY","index":"5"},{"bias":0.06477448349804442,"type":"hidden","squash":"RELU","index":"6"},{"bias":-3.1021366685789973,"type":"output","squash":"RELU","index":"7"}],"connections":[{"weight":1,"from":"5","to":"5","gater":6},{"weight":1,"from":"6","to":"6","gater":null},{"weight":-0.00585266827152324,"from":0,"to":5,"gater":null},{"weight":-0.04714111617289185,"from":1,"to":5,"gater":null},{"weight":0.0827437010660865,"from":2,"to":5,"gater":null},{"weight":-0.794563445679669,"from":0,"to":7,"gater":null},{"weight":0.03128007567497146,"from":3,"to":5,"gater":null},{"weight":0.023534840247686345,"from":1,"to":7,"gater":3},{"weight":-0.030584405719202665,"from":4,"to":5,"gater":null},{"weight":-0.046412579401978515,"from":2,"to":7,"gater":2},{"weight":-0.09529622104686171,"from":3,"to":7,"gater":null},{"weight":0.04095217250657379,"from":6,"to":5,"gater":null},{"weight":0.06933036853010713,"from":4,"to":7,"gater":null},{"weight":0.1511437332387351,"from":5,"to":7,"gater":null},{"weight":-0.028524771915109964,"from":0,"to":6,"gater":null},{"weight":-0.046412579401978515,"from":2,"to":6,"gater":2},{"weight":-0.09529622104686171,"from":3,"to":6,"gater":null},{"weight":-0.012785664769956864,"from":4,"to":6,"gater":null},{"weight":-0.08522174805526968,"from":5,"to":6,"gater":null},{"weight":-0.0062336182902426734,"from":7,"to":5,"gater":null},{"weight":0.03729470347555722,"from":7,"to":6,"gater":null},{"weight":0.0042867867849913505,"from":6,"to":7,"gater":null}],"input":5,"output":1}`

**Comment by wagenaartje** [comment]
I just tried your 3 networks provided with the given input and they all seem to output fine (https://jsfiddle.net/wagenaartje/r4j21oer/). If you keep on getting issues, try to remove the following activation functions:

```js
Activation.SOFTSIGN
Activation.SINUSOID
Activation.BENT_IDENTITY
Activation.ABSOLUTE
```

Maybe that'll fix it.

**Comment by zammbi** [comment]
I've been currently running:
        Methods.Activation.TANH,
        Methods.Activation.SOFTSIGN,
        Methods.Activation.SINUSOID,
        Methods.Activation.BIPOLAR_SIGMOID,

So far no issues. When I was getting into a NaN/Infinity state, it couldn't get out of it.

I've just changed my code to avoid small numbers. But I'll see if I can get a repeatable case using your jsfiddle code.

**Comment by wagenaartje** [comment]
Hmm. I'll do some random debugging tasks to see if I'm getting any `NaN/Infinity` values.

**Comment by zammbi** [comment]
I can't repeat it in a jsFiddle case. It might be because the numbers being reported by the debugger are not exactly those numbers...
https://jsfiddle.net/38hkyfv7/3/

**Comment by zammbi** [comment]
I narrowed it down to these causing it:

Methods.Activation.RELU
Methods.Activation.IDENTITY
Methods.Activation.ABSOLUTE
Methods.Activation.BENT_IDENTITY

I ran each activation, one at a time.

**Comment by wagenaartje** [comment]
That seems to match exactly the list I put up here https://github.com/wagenaartje/neataptic/issues/13. The only difference is that I removed SOFTPLUS some time ago.

Each of your named funtions produce large values, they are the source of the largue values, but not the source of NaN/Infinity.

This is a problem for which i'm not sure there is a fix. It's a result of very high numbers internally.  I've got a stackexchange question up here. https://stats.stackexchange.com/questions/277618/neural-network-failing-because-of-infinity

**Comment by wagenaartje** [comment]
I did, but i'll try some more. Even the big players like Keras only have a limited amount of activation functions https://keras.io/activations/.

**Comment by zammbi** [comment]
They have 13 different ones in Keras from what I can see: softmax, elu, softplus, softsign, relu, tanh, sigmoid, hard_sigmoid, linear, LeakyReLU, PReLU, ELU, ThresholdedReLU.

**Comment by wagenaartje** [comment]
TANH should definitely not be an issue. It's has a normalised output and will never output `NaN` (only when given `NaN`) or `Infinity`, it even accepts `Infinity` as input.

They have 13 'different' ones, but 5 of them are (rectified) linear units. You named elu twice by the way.

But i'm going to do some random mutations now and see with which settings im getting `NaN` outputs. `Infinity` will only be given as output if the output neuron's activation function is not normalised. 

However, I have already set up quite some NEAT algorithms and I've only encountered this issue once. So it could be very specific to your scoring function. But i'm looking in to it now.

**Comment by zammbi** [comment]
"TANH should definitely not be an issue."
Oh whoops, I used the wrong list. 

I mean I used RELU which had issues and they use it too.

"You named elu twice by the way"
They have two versions of it.

**Comment by wagenaartje** [comment]
Yes RELU could cause the issue by outputting high values. I just created some complete network randomizer that mutates + crossovers networks: https://jsfiddle.net/wagenaartje/81Lgaj0e/. The problem is not in mutation and crossovering I noticed.

_However_, when I try to do some backprop, NaN is outputted. So i'm trying to find the cause of this right now. https://jsfiddle.net/wagenaartje/81Lgaj0e/1/

**Comment by wagenaartje** [comment]
So I did quite some debugging. I know what the problem is, but I don't know what the solution is.

Basically I set up a lot of checkers in `node.js` to debug, e.g.:

```js
if(isNaN(this.state)){console.log('State', this.squash.name, this.connections.self.gain, this.connections.self.weight, this.state, this.bias); throw new Error('NaN');}
if(this.state == Infinity || this.state == -Infinity){console.log('State', this.squash.name, this.connections.self.gain, this.connections.self.weight, this.state, this.bias); throw new Error('Infinity');}
if(this.state > 1e2){console.log('State', this.squash.name, this.connections.self.gain, this.connections.self.weight, this.state, this.bias); throw new Error('Large output');}
```

I added the last of those 3 checks afterwards. Never do I get a `NaN` or `Infinity` debug error, I always get the `Large output` error. So basically what is happening: activation functions without a fixed range build up and up until they reach `Infinity`, and then the network errors.

So I took a look at keras/tensorflow:

* https://github.com/fchollet/keras/blob/562860ca42d8dd65e87e3f0b8e76b811a718a9a5/keras/activations.py
* https://github.com/fchollet/keras/blob/3061fcce608677578d2db72203171f9a42badb2a/keras/backend/tensorflow_backend.py
* https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/ops

But I really couldn't find anything. I thought of limiting weights/biases, but I think this would disturb the NEAT algorithm.

The problem is _not_ specificely backprop related. I also activated a random neural network up to 12000 times and it errors within 1000 iterations most of the times. The issue is just that activations just keeping building up.

However, removing selfconnections from the network actually clears the error for activations. It also reduces the errors on backpropagation. 

<hr>

The backprop algorithm is modified from Synaptic, and I encountered a [weird bug](https://github.com/cazala/synaptic/issues/224) . However I noticed one other thing:



**The self connection weight is always 1**, making this weight dynamic might solve the problems. But I shall look into it more.

**Comment by zammbi** [comment]
"The backprop algorithm is modified from Synaptic, and I encountered a weird bug"
I also noticed this. I just thought I didn't know enough of what was happening (it was the first example I ran).

**Comment by zammbi** [comment]
I think I figured it out why I was having so much trouble and now longer when I simplified my model. 

I was doing this `network.score-= number;`. This was causing NaNs to happen. But no longer an issue if I checked `network.score` was undefined and set it to 0 when it is.

**Comment by wagenaartje** [comment]
I checked out your JSFiddle. The `NaN` problem in that fiddle was not caused by Neataptic. It was caused by the fact that `genome.score` is never initialised in Neataptic. So when you perform:

```javascript
genome.score -= Math.abs(answer - (var1 * var2 * var3));
```

it's basically

```javascript
undefined -= score;
```

which will always result in `NaN`. You just forget to initialise `genome.score` in your code.

I know you made this script pure for showing the error, but I've got some comments

* I don't think you're using `const` correctly, use `var` instead
* You never set the genome's score to `0`, when you use elitism, scores of those genomes will be kept. 

<hr>

On the side note, you have made me include `genome.score = 0` in the source code of `genome.evalute()`. A commit will be pushed soon.

**Comment by zammbi** [comment]
Thanks for pointing out my mistakes out. I'm used to nicer languages with better tools :)

"You just forget to initialise genome.score in your code."
Yeah, but shouldn't Neataptic's default score be 0 instead of undefined? 

"I don't think you're using const correctly, use var instead"
Seems right to me?
https://medium.com/javascript-scene/javascript-es6-var-let-or-const-ba58b8dcde75

"when you use elitism, scores of those genomes will be kept"
In this case, would it be better to reset these scores?

**Comment by wagenaartje** [comment]
"Yeah, but shouldn't Neataptic's default score be 0 instead of undefined?"
That is hard to say. Because for some GA's, 0 might be large score, while for some, it will be the lowest score. And the `NEAT` instance relies on a fitness function, so it calculates the score as `genome.score = fitnessFunction(genome)`. So the score will always be set. The `undefined` will only cause a problem when implementing the `NEAT` instance in the Agar.io-style way.

"Seems right to me?"
You're right. It's possible, but it just seemed really weird for me to use `const` for a variable that has a different value every iteration :)

"when you use elitism, scores of those genomes will be kept"
Definitely! Implementing it now.

**Comment by wagenaartje** [comment]
https://github.com/wagenaartje/neataptic/commit/6b3a2cd8828d97fa92463ffe492be43297339caa - scores reset to `null` now
